# Introduction

The goal of this project is to analyze non-normal distributed data of open source GitHub projects from a dataset provided by .

The project should show an ability to explore data, handle ambiguity, and weave insights into a narrative. It should have well-formed observations and understanding of the nuances of the data.

In [1]:
# Import required modules
import pandas as pd
import requests
import time
import matplotlib.pyplot as plt
from scipy import stats
import numpy as np
import seaborn as sns
from datetime import timedelta

# Data Extraction

The dataset is provided from two files "usage.csv" and "projects.csv". They are selected usage records of open source GitHub projects.

In [2]:
usage = pd.read_csv("usage.csv")

# To view initial structure of dataset
usage.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 566094 entries, 0 to 566093
Data columns (total 6 columns):
id            566094 non-null object
actor_id      444652 non-null object
project_id    566094 non-null object
account_id    566094 non-null object
started_at    566094 non-null object
ended_at      566094 non-null object
dtypes: object(6)
memory usage: 25.9+ MB


There are 566094 observations/rows in usage and 6 columns/variables. Every column, except for the actor_id has non-null values.

In [3]:
project = pd.read_csv("projects.csv")
project.columns=['project_id','project_size']
project.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12907 entries, 0 to 12906
Data columns (total 2 columns):
project_id      12907 non-null object
project_size    12906 non-null float64
dtypes: float64(1), object(1)
memory usage: 201.8+ KB


The initial view of the projects data file shows that there is 1 null value.

In [4]:
# Merging two datasets
merged = pd.merge(usage, project, on='project_id', how='outer')

In [5]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 566094 entries, 0 to 566093
Data columns (total 7 columns):
id              566094 non-null object
actor_id        444652 non-null object
project_id      566094 non-null object
account_id      566094 non-null object
started_at      566094 non-null object
ended_at        566094 non-null object
project_size    563510 non-null float64
dtypes: float64(1), object(6)
memory usage: 34.6+ MB


# Data Wrangling/Munging

In [6]:
merged['started_at'] = pd.to_timedelta('00:' + merged['started_at'], errors='coerce') 
merged['ended_at'] = pd.to_timedelta('00:' + merged['ended_at'], errors='coerce')   

Due to the non-powerful nature of my tablet-laptop, I have subsetted the data to 384 observations. This number was determined by calculating the sample size with a population of 566,094 (the total number of observations), a confidence level of 95%, and a margin of error of 5%. Since I am limiting the dataset, I also removed all null values.

In [7]:
no_na = merged[pd.notnull(merged['actor_id'])]
no_na = no_na[pd.notnull(merged['project_size'])]
small = no_na.sample(384)
small = small.reset_index(drop=True)

C:\Users\Kevin\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


In [10]:
small.head()

id  \
0  bf8cab9f9149bab845afe66325414292701b11c61861ff...   
1  41646c5e9b8a975b2f8640a0676c285fe4874752210a69...   
2  382bf302eb86c807626e77dd6ea29299fdbf496a0d1d1f...   
3  4214f67e79d34ad0cea7be77bae0ab9900f59650718c09...   
4  fed1962fa12855cf01eeaee25a726d589d810a3326c9ab...   

                                            actor_id  \
0  c46477d6f6c30a8beb1dfded05831e088eca300781df6d...   
1  c77914feecb11bfea419d572c4ff7c2576d32a206f5221...   
2  779d9b3f3bb0396df4adc385bc8de0487dd32b4acb55c6...   
3  467a3f755fafd35efc28ff5ebacd193c61c5e9ba388a84...   
4  25aab5baedbe1b53d5312f713dfa32bcc18f944af914d9...   

                                          project_id  \
0  da4499ef83cdf572787f0e1ae3e81684116a3e296be777...   
1  0f82fe8166494df390f38385cd2d1691a4374fe1c0d6c7...   
2  d452291abaae5727935e6c80768bb51fd8ceaf6c537dd1...   
3  f09396375c9a706d291c6264cdcfd441182d9d422f2b7d...   
4  79c6bbae98854c8f2fac9b78411575db8a922af71dcb24...   

                                          account_id      started_at  \
0  036eadeac583ba630ec8d7a22a3b495d2f9e9268c28e0f... 00:42:31.300000   
1  09fdd9137c9f5885d4a413864e9839c5d98fefb9588715... 00:40:53.600000   
2  203ea8da668de8d4cd68cfe51dcf998bbc7bea8b1b7206... 00:01:12.700000   
3  d2fefe1774b77114507fdce4a9c63cf13a6bede7afff9d... 00:57:02.500000   
4  1e9deba2f445989233c8397686ece4c92c93d848a943f5... 00:14:33.900000   

         ended_at  project_size  
0        01:35:22        7764.0  
1 00:44:01.900000      192729.0  
2 00:01:18.700000          29.0  
3 00:58:00.200000          71.0  
4 00:15:14.100000         114.0

Since the imported data only measured time in minutes and dates, events such as the start time being after the end time occured. This is not realisticall possible. Therefore the next code segment checks that the laws of time have not been broken, if so, it makes adds an hour until the end date is after the start date. This results in a more accurate and realistic time of, for example 2 minutes, than 23 hours and 58 minutes.

In [9]:
i=0
while i < len(small):
    try:
        while small['ended_at'][i]<small['started_at'][i]:
            small['ended_at'][i] = small['ended_at'][i]+timedelta(0,60*60)
    except TypeError:
        pass
    i += 1

C:\Users\Kevin\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [11]:
# Calculate time_of_usage

calc = small.apply(lambda row: row['ended_at'] - row['started_at'], axis = 1)

small['time_of_usage'] = calc

# Check / Test
small

id  \
0    bf8cab9f9149bab845afe66325414292701b11c61861ff...   
1    41646c5e9b8a975b2f8640a0676c285fe4874752210a69...   
2    382bf302eb86c807626e77dd6ea29299fdbf496a0d1d1f...   
3    4214f67e79d34ad0cea7be77bae0ab9900f59650718c09...   
4    fed1962fa12855cf01eeaee25a726d589d810a3326c9ab...   
5    3769fe3cab948de7f4fedc5dad1828c080cfc5eeed209e...   
6    fd556e301ec5c4f650d7d7cf9db5f5c38c5f96a90d581a...   
7    dd803a121e1f906efda6e85ae7ca51ba9acf9315ff3075...   
8    0e7af1c84ab57e2adb9cf8fdce0da4a5b36881719fba80...   
9    9347e5b962f1142591129ccfc16c7cb6bd0939e7cd6ab8...   
10   011440c4d86cf77b6042d99e2e4e78bba146c9db64fd79...   
11   3bdd5c7e089ebf85ced67967fa5f575ae396a337bcdfaa...   
12   45ac4069c617b42c08e24367f765ef5a16e0b8cd14b651...   
13   ca926c72d2ed5d3c2acc777e37edb4e8745d5ee4202fc4...   
14   06d2f5694e4cb19cf48a710ecf3ae25147d6ee5bfe284d...   
15   8f10f22615ea75249642f885bbf93469ab066c5e1df1af...   
16   2a85d1445521a2a94fa3c469c6f6a0171d509cc8197f6e...   
17   95b13e0230974873e9bb6c618b6d0f06e3ae572a4c8706...   
18   f7351fb0ffc789b52d4807f31960cb209a9125cda1c2a9...   
19   d124945baf389de595cbc7de3014793b07676c71f4bb25...   
20   4051185bc7ab07a5346b872386857a33c0d6b33a81f0c2...   
21   5cdcfd7d5703777f56e11d85dbefc81b770ce5fe4c4d45...   
22   40af9515906a37959824f6f42a888f73c92858e4babd70...   
23   51bec08c744424fcab4c7d962daa342090bda6586a5267...   
24   a6a8515f222b31d9b2cf0baff5a29804f761aaf358afb7...   
25   1e3914c3ae204f8dcd7aedb12e64c95efef06f083b6e75...   
26   86ea073d9c56e00cd9eb5b2b6c01f92a11b7065d6b0618...   
27   d5064a2a87d42c6a50628170f9f8f8829f0501e414607f...   
28   7dc41d73cf0198f201dbd40e2eb209cf3bd4520c2b85dd...   
29   f41de53f5507d58e2cb0169d84dd7586c055eb521d6834...   
..                                                 ...   
354  e6f9b5b09266e2c1f6f4a4e7e46b35b7bd80ac4fcd206a...   
355  83dac1a9d93399ed90dd60f2b078e662c6402ce708bd88...   
356  986dfcee5a19015fb06defda57f7f74af42c8ac9406c71...   
357  8601601edb63f106c7daaa16b0055c4c61ae183998f06a...   
358  06a13aad7db58248125e2b9f06e3efaa8c38de697d3171...   
359  62cf7d0fb3c06c7973dd1da437679ec440cfe089abe802...   
360  01ff4cd361fce2b4117c05f16a1096d9b3a87b552fb116...   
361  8de2e67f3d14140175c2608ce8c6567dd637568b0868cc...   
362  c002dbed658b4036b8d17b7aca013cd503ee99649a1870...   
363  f6fa1f069027b5b0bdc0b773190f725564d5a3ce8a1e49...   
364  56d06b2385441ecd07d874e1ce633e4286a85a1c574732...   
365  352abe274157a083ae1cdbfb536e37dc1ec459f04d72a8...   
366  429146624edcf0eb680576e9bc4f3d09c495ebd9b28715...   
367  adc0c7f7419640f5ba347c2a197a670fdb55c4fc913a83...   
368  a93ac8ebd385bd3f1becd41faf16f54f824d5227445313...   
369  e76acebed54aa177bb610b5674549a9cf6c79828eeadaa...   
370  86fb612c746980d99c985abb706168fb82b048964a34f8...   
371  e098f34d35a5af654c2976054d70c15bca814d4f52015f...   
372  e2f657d20a1607b840f14241ac837efcc7e36ea3e9c3fe...   
373  6350cc2a9d4b1b8a254412a92078cdcf5736ec0b612c46...   
374  4d2ca0b2ab05520458203cb2a995560a7dc9db86e2f276...   
375  1335a69aa41eed0e440c64d04c3518426b56c301470f94...   
376  8b083b195550038b705ae274cbdf1499573a8ac195a2bf...   
377  9f7936d343b9b5a6e3625dc3a25e1cd533d731b851c402...   
378  b5ae62f752c9d5b9a6115ed078ad2d3c2484681f8a77a4...   
379  4852d58ce0f194b546aab3ec614e5221d3b538ca24ad27...   
380  adbb558b7a42aa5842d097115633f32978ed9cdbf46322...   
381  ae8bda500b738e77400ba2233a790d2617f50e114f8029...   
382  41efa109f1311db9ab349b878e8d5f8739527d765c3ac2...   
383  501090f9fcc5d6e4f741e0a14eb48eecf67aa146e40fee...   

                                              actor_id  \
0    c46477d6f6c30a8beb1dfded05831e088eca300781df6d...   
1    c77914feecb11bfea419d572c4ff7c2576d32a206f5221...   
2    779d9b3f3bb0396df4adc385bc8de0487dd32b4acb55c6...   
3    467a3f755fafd35efc28ff5ebacd193c61c5e9ba388a84...   
4    25aab5baedbe1b53d5312f713dfa32bcc18f944af914d9...   
5    e9d3a9cb09cea8812bd1d93cffa6bfbe700f8afe32c482...   
6    4a80bd9ca1525c36321ad59fba564c65eeb286cfc32c

In [ ]:
merged

# Section 1 Questions

#1 Identify the top 10 actors - by total usage, by accounts built against, and by projects built against

In [ ]:
# By Total Usage

In [15]:
# By Acccounts Built Against
small.groupby(by=['actor_id']).sum()

project_size
actor_id                                                        
00bdf40f70f54ea4a464dc2f34b853f9c78e70cd55affb6...       23729.0
017a7c8cf1340e89f3ec23e902606430bd911804bbb3c38...       60907.0
04477b6a449493f7d060508eecdf42c20f6f0f7f7fb7ee9...       20376.0
059b10950ee999efb6a8b69308586198994871484c92c5a...      169312.0
062ec79a5705b77e62504dfdb8e171731ebd1db9499a9e7...          26.0
069c778722156b659771450cb6a92987e6decc2dd324e6a...         111.0
074da42d4da865db011f49ad58cc40e24aa2b2aa2e5bb1c...       49172.0
0756faeebacbd0f8e000815bef3554014bc1bf0ffcb38e7...           3.0
09a31247440b37b5db2f82a156138395ee872eac459f251...         165.0
09ce4ea439d3e17813f5f861481984c9046ab25c15cd703...        2235.0
0a300b3d571ce2f69820e5bed369a7bbdecfbb2c7c55d6d...        3093.0
0a40a3f330a55ecc336a9019edd5a57df118e2ae52495f0...        3582.0
0c4a7cb2dc809682ec30ae6d015024693b6ba5be14a4fae...        3518.0
0d70dd5f1f77060d765fa9e5de5a037323b06462960d1fe...       15400.0
0d87b6bdc38b23289520d89a7478d1a7513b198cbd8d53c...          31.0
0e13cd2004f8ed58966b9b65da082d1e242f558a62ac322...          46.0
0e77438cfafd2c1f558c84c7f6faae4e7b7a0142eb1a8b8...       11897.0
0f570b6adc0eb632ea94b74736821b49d2a9f49a44d3966...      215835.0
0f876d3b2f06e7a4c485aeaf00d9e03004b2ce8d390b49e...         108.0
0f979867d6ace88fdc3928a831ce0d5916bde3c371006f4...         476.0
11103480a4950424a22480373744fda742cbed162d43bf1...       49172.0
12631cb1b00605e2478c12f0873c7cfb24e0b816ee2af3a...          87.0
15d4bbc943868d9ff07fc1887a0a63863d411407bc8b148...        2069.0
15e7fb008ebd4584adc1fd09f43476b38c88af1c21f1872...      225211.0
18b1f6d3fc1661352c1ee6f7b35e85f445e72899d6d6f82...       80898.0
193b7b25ed225715abfa2cbb421ced3d898c1efee144213...       45972.0
1b0a011a68ea26c8eba60dbe87a80b8f5711f8521bb8fb8...         601.0
1ba620584b308be80712ff4f3a735582e3cc13afcf090c6...       62610.0
1c78b8f9433481c97c28ee17923bb9a81138727abb8c054...       44539.0
1c85940b8e26306f9730dcc43d56264cbee5997d11ccabf...       68578.0
...                                                          ...
e0e1a82f3998903eb81d68c03023ffbff3e784f20fe994f...       10304.0
e1cdcc3b1211ff1cca254593ed6eeb213eea2ae74cc74f6...         184.0
e2b273ec6ec6bb736e9ec3f7fef5bf4e626496d73b3e29b...       23352.0
e37ac3d1c201a0bdad795c6f329a4dd6cac38db546ec87e...        3949.0
e38e49c4d2e5704392e754c6e8657e6a4d3346862b500fb...       23351.0
e584bc8d189e0bd6cd8bf361343bd94eed51e75eb99de13...         760.0
e5898be21a673ee7a5b3597db3dc3be2be6eb8eda79348a...      547188.0
e658866173fb4351e1684d2473d82f79d4ff1e9ec11771c...         419.0
e6f4e16e196d4942304e2fcff240d5239e1b5e1433a5659...       60572.0
e9d3a9cb09cea8812bd1d93cffa6bfbe700f8afe32c482b...         404.0
e9fc68dc702a0f6fa11d32de4e4c7f8ca1d56003aee6b66...       44539.0
eb12a3e7eeafc50b0642b74eb315dd257e609052d76b89f...       88882.0
eb618f8d96e16009d808e9cbe994a5b5b1b2ef3cee7c3a2...          67.0
ee16415227dc8c75613227f08c64144403621ae586a7a61...        6110.0
efa243c2b1f4f431ac5d06373f6ec056ea262a4b369fced...       47804.0
f0fe17a2d64ecc1c99169cd0fd4a7e6236f5324044529cf...        5753.0
f2049c8a1546d1689f61357f4ba2715e1301bb28a6f0d2f...       30741.0
f2d1786de7a5424a6b08212b495bee4a1e43fb078c1cb67...        1729.0
f2ea066a473a22a2dbf55b69d9790778950f2c5007390f7...         267.0
f678d90f2fc5207ab2b61bcf394d21ce1aa0f639aed6ec7...         196.0
f88866d7e414b5fd4fe2875eaee1706f0fbac082e006bc7...        9353.0
f9fad860610266fe7c95ee663b04e801a9d242589748e72...      202981.0
fa61c72ad2116227457b530850296edae67ba9547a4e860...         423.0
fb08f50a60c130badd712a225f4557e02ee6ddefdafea2e...        1687.0
fc186e39d153e63b1135927e75eeb0d94b2e249d1f5949a...        4250.0
fc360ffb3458f7186a1d8661ec30691a8cfdfb0b70a4fa5...       60572.0
fd3b816e27d24e0a26303ed4e1eb1a626ca4ad0bb28a307...      145158.0
fda2a0cd2388b39622a81a164aa43b69e218525d85aa56d...        2060.0
fe4cf359b7bcf25f8c18f53bef360742ba2f7baff0013a9...       32927.0
ff1e72acad9a1620b6905e

In [16]:
# By Projects Built Against

#2 For each ranking, also provide the scalar amounts

#3 Provide a five-number summary of usage for each week, and a five-number summary of total usage per account for each week

In [ ]:
# The provided datasets do not specify weeks so it is not possible to know if these events
# happened on different weeks or the same week

merged.describe()

#4 Provide some comments on the following: Where can a t-test be applied to these collections? Provide some examples of where it can and cannot, insights on why, and other implications

In [ ]:
A t-test can be us

#5 Provide an analysis that shows the relationship between the following variables, along with insights drawn out and comments on your choices:
- Total usage per account
- Unique Actors per account
- Two other variables

In [ ]:
rho, pval = stats.spearmanr(small)

pvals = pd.DataFrame(pval)
#pvals = pvals.rename(index={0: 'outlook',1: 'net_income',2: 'ft_employee',3: 'div_yield',4: 'five_year',5: 'ten_year',6: 'alcohol',7: 'gambling',8: 'tobacco',9: 'controversial_weapons',10: 'small_arms',11: 'military_contracting',12: 'coal',13: 'dont_use',14: 'dont_trust',15: 'assets',16: 'market_value',17: 'revenue',18: 'net_per_emp'})
#round(pvals, 3)
pval

In [ ]:
# Fill in blanks
merged2["ft_employee"].fillna(merged2["employees"], inplace=True)
merged2["sector"].fillna(merged2["sector_f"], inplace=True)
merged2["industry"].fillna(merged2["industry_f"], inplace=True)

# Removing Duplicate Column
merged2 = merged2.drop('employees', 1)
merged2 = merged2.drop('sector_f', 1)
merged2 = merged2.drop('industry_f', 1)

In [ ]:
# Calculate Net Income Per an Employee

calc = merged2.apply(lambda row: row['net_income'] / row['ft_employee'], axis = 1)

merged2['net_per_emp'] = calc

# Check / Test
merged2

In [ ]:
# Correlation and P-Values

matrix = calcdf
matrix = matrix.drop('date',1)
matrix = matrix.drop('ticker',1)
matrix = matrix.drop('security',1)
matrix = matrix.drop('sector',1)
matrix = matrix.drop('industry',1)
matrix = matrix.drop('adult',1)

rho, pval = stats.spearmanr(matrix)
# print('stats.spearmanr - cor:\n', rho)
# print('stats.spearmanr - pval\n', pval)

#pval

# round(pd.DataFrame(pval), 3)

pvals = pd.DataFrame(pval, columns=['outlook','net_income','ft_employee','div_yield','five_year','ten_year','alcohol','gambling','tobacco','controversial_weapons','small_arms','military_contracting','coal','dont_use','dont_trust','assets','market_value','revenue','net_per_emp'])
pvals = pvals.rename(index={0: 'outlook',1: 'net_income',2: 'ft_employee',3: 'div_yield',4: 'five_year',5: 'ten_year',6: 'alcohol',7: 'gambling',8: 'tobacco',9: 'controversial_weapons',10: 'small_arms',11: 'military_contracting',12: 'coal',13: 'dont_use',14: 'dont_trust',15: 'assets',16: 'market_value',17: 'revenue',18: 'net_per_emp'})
round(pvals, 3)

In [ ]:
mask = np.zeros_like(pvals)
mask[np.triu_indices_from(mask)] = True
with sns.axes_style("white"):
    p2 = sns.heatmap(pvals, mask=mask, square=True, cmap="PuOr")

# sns.heatmap(pvals, square=True)

In [ ]:
# Not normally distributed so Spearman
cm = matrix.corr(method='spearman')
cm

In [ ]:
sns.heatmap(cm, square=True)
plt.yticks(rotation=0)
plt.xticks(rotation=90)

In [ ]:
items = ['outlook','net_income','ft_employee','div_yield','five_year','ten_year','alcohol','gambling','tobacco','controversial_weapons','small_arms','military_contracting','coal','dont_use','dont_trust','assets','market_value','revenue','net_per_emp']
arr = []

for i in items:
    for j in items:
        if pvals.loc[i,j] <= 0.05 and (cm.loc[i,j] > 0.5 or cm.loc[i,j] < -0.5) and i != j and [j,i, pvals.loc[j,i], cm.loc[j,i]] not in arr:
            arr.append([i,j, pvals.loc[i,j], cm.loc[i,j]])
        

arr = pd.DataFrame(arr,columns=['1','2','P-Value','R'])
arr.sort_values('R', ascending = True)

In [ ]:
five_year_df = calcdf[calcdf.outlook.notnull()]
five_year_df = five_year_df[five_year_df.five_year.notnull()]
five_year_df

x, y = five_year_df.outlook, five_year_df.five_year
slope, intercept, r_value, p_value, std_err = stats.linregress(x,y)
line = x*slope + intercept
five_corr = plt.plot(x,y, 'x', x, line, color = 'black')
plt.xlabel('% Positive Business Outlook')
plt.ylabel('% Return on Investment')
plt.show()

print('Line Equation: Five Year Return = ({})*Outlook +{}'.format(slope,intercept))
print('Line of Best Fit Correlation: {}'.format(r_value))

In [ ]:
ten_year_df = calcdf[calcdf.outlook.notnull()]
ten_year_df = ten_year_df[ten_year_df.ten_year.notnull()]
ten_year_df

x, y = ten_year_df.outlook, ten_year_df.five_year
slope, intercept, r_value, p_value, std_err = stats.linregress(x,y)
line = x*slope + intercept
five_corr = plt.plot(x,y, 'x', x, line, color = 'black')
plt.xlabel('% Positive Business Outlook')
plt.ylabel('% Return on Investment')
plt.show()

print('Line Equation: Ten Year Return = ({})*Outlook +{}'.format(slope,intercept))
print('Line of Best Fit Correlation: {}'.format(r_value))

In future data collections, the information will be gathered but the date will change. Thus, it will be possible to develop a time series forcast.

In [ ]:
X = calcdf[np.isfinite(calcdf['outlook'])]
X = X[np.isfinite(X['net_income'])]
X = X[np.isfinite(X['ft_employee'])]
X = X[np.isfinite(X['div_yield'])]
X = X[np.isfinite(X['five_year'])]
X = X[np.isfinite(X['assets'])]
X = X[np.isfinite(X['market_value'])]
X = X[np.isfinite(X['revenue'])]

Y = np.array(X['five_year'])
X = np.array(X.drop(['ticker','security','sector','industry','alcohol','adult','gambling','tobacco','controversial_weapons',
                          'small_arms','military_contracting','coal','dont_use','dont_trust','five_year','ten_year'],1))

In [ ]:
from sklearn import preprocessing, cross_validation
from sklearn.linear_model import LinearRegression

In [ ]:
X = preprocessing.scale(X)

In [ ]:
X

In [ ]:
# Checking Lengths Match
print(len(X), len(Y))

X_train, X_test, Y_train, Y_test = cross_validation.train_test_split(X, Y, test_size=0.2)

clf = LinearRegression(n_jobs = -1)
clf.fit(X_train, Y_train)

squarederror = clf.score(X_test, Y_test)

print(squarederror)

# Predicted
forecast = clf.predict(X)

print(forecast)